src/Protocols/DfLike.hs

```diff
@@ -41,931 +68,1546 @@ module Protocols.DfLike
   , bundleVec
   , unbundleVec
   , roundrobin
-  , CollectMode(NoSkip, Skip, Parallel)
+  , Df.CollectMode(..)
   , roundrobinCollect
   , registerFwd
   , registerBwd
+  , fifo
 
     -- * Simulation functions
   , drive
-  , stall
   , sample
+  , stall
   , simulate
-
-    -- * Internals
-  , forceResetSanity
   ) where
 
--- base
+
 import qualified Prelude as P
-import           Prelude hiding
-  (map, const, fst, snd, pure, either, filter, zipWith, zip)
-import           Control.Applicative (Alternative ((<|>)))
-import           Data.Bool (bool)
+import           Control.Arrow ((***))
+import           Control.Monad (when)
+import           Control.Monad.State (State, runState, get, put)
+import           Clash.Prelude hiding
+                   (map, fst, snd, zipWith, const, pure, filter, either, zip, select, sample, simulate)
+import qualified Clash.Prelude as C
 import qualified Data.Bifunctor as B
-import           Data.Bifunctor (Bifunctor)
-import           Data.Kind (Type)
-import qualified Data.Maybe as Maybe
+import           Data.Maybe (fromMaybe, isNothing)
 import           Data.Proxy (Proxy(..))
-import qualified Data.List.NonEmpty
-import           Data.List ((\\))
+import           Data.Tuple (swap)
 import           GHC.Stack (HasCallStack)
 
 -- me
-import           Protocols.Internal hiding (Ack(..))
+import           Protocols.Axi4.Common
+import           Protocols.Axi4.ReadAddress
+import           Protocols.Axi4.ReadData
+import           Protocols.Axi4.WriteAddress
+import           Protocols.Axi4.WriteData
+import           Protocols.Axi4.WriteResponse
+import           Protocols.Df (Data(..), Df)
+import qualified Protocols.Df as Df
+import           Protocols.Internal
+
+
+-- | Class for protocols that are "similar" to 'Df',
+-- i.e. they can be converted into a 'Df' port using a 'Circuit' (see 'toDfCircuit').
+-- This is for protocols that carry some "interesting" data,
+-- as well as some "uninteresting" data (e.g. address, burst length).
+-- The 'Circuit' should abstract away the complexities of each protocol,
+-- so that they can be dealt with uniformly using 'Df'.
+-- For pipelined protocols, which can carry both in the same cycle,
+-- the 'Circuit' should pass along the interesting parts
+-- but not the uninteresting parts.
+--
+-- Can take parameters, e.g. for addresses.
+-- Defaults to being the right side of the circuit (@Fwd ~ Otp, Bwd ~ Inp@),
+-- but this can be switched using 'Reverse' from 'Protocols.Internal'.
+-- Supports both bwd (input/read) data and fwd (output/write) data.
+class (Protocol df) => DfLike df where
```

Comment:
Perhaps we should rename this `ToDf` or `DfConv`? It's not really that the protocols have to be anything like `Df`, the only requirement is that they can be converted to and from it.